gst/nnstreamer/tensor_query/tensor_query_common.c

```diff
@@ -60,32 +57,422 @@ typedef struct
+          return -EIO;
+        }
+        return 0;
+      } else {
+        /* receive data_info */
+        if (!query_tcp_receive (conn->socket, (uint8_t *) & data->data_info,
+                sizeof (TensorQueryDataInfo), conn->cancellable)) {
+          nns_loge ("Failed to receive from socket");
+          return -EIO;
+        }
+      }
+    }
+      break;
+    default:
+      /* NYI */
+      return -ENOSYS;
+  }
+  return 0;
+}
+
+/**
+ * @brief send command to connected device.
+ * @return 0 if OK, negative value if error
+ */
+int
+nnstreamer_query_send (query_connection_handle connection,
+    TensorQueryCommandData * data, uint32_t timeout_ms)
+{
+  TensorQueryConnection *conn = (TensorQueryConnection *) connection;
+  UNUSED (timeout_ms);
+  if (!data) {
+    nns_loge ("Sending data is NULL");
+    return -EINVAL;
+  }
+  if (!conn) {
+    nns_loge ("Invalid connection data");
+    return -EINVAL;
+  }
+
+  switch (conn->protocol) {
+    case _TENSOR_QUERY_PROTOCOL_TCP:
+      if (!query_tcp_send (conn->socket, (uint8_t *) & data->cmd,
+              sizeof (TensorQueryCommand), conn->cancellable)) {
+        nns_loge ("Failed to send to socket");
+        return -EIO;
+      }
+      if (data->cmd == _TENSOR_QUERY_CMD_TRANSFER_DATA) {
+        /* send data */
+        if (!query_tcp_send (conn->socket, (uint8_t *) data->data.data,
+                data->data.size, conn->cancellable)) {
+          nns_loge ("Failed to send to socket");
+          return -EIO;
+        }
+      } else {
+        /* send data_info */
+        if (!query_tcp_send (conn->socket, (uint8_t *) & data->data_info,
+                sizeof (TensorQueryDataInfo), conn->cancellable)) {
+          nns_loge ("Failed to send to socket");
+          return -EIO;
+        }
+      }
+      break;
+    default:
+      /* NYI */
+      return -ENOSYS;
+  }
+  return 0;
+}
+
+/**
+ * @brief free connection
+ * @return 0 if OK, negative value if error
+ */
+int
+nnstreamer_query_close (query_connection_handle connection)
+{
+  TensorQueryConnection *conn = (TensorQueryConnection *) connection;
+  if (!conn) {
+    nns_loge ("Invalid connection data");
+    return -EINVAL;
+  }
+  switch (conn->protocol) {
+    case _TENSOR_QUERY_PROTOCOL_TCP:
+    {
+      GError *err;
+      if (!g_socket_close (conn->socket, &err)) {
+        nns_loge ("Failed to close socket: %s", err->message);
+        g_error_free (err);
+        return -EIO;
+      }
+      g_object_unref (conn->socket);
+      g_object_unref (conn->cancellable);
+    }
+      break;
+    default:
+      /* NYI */
+      return -ENOSYS;
+  }
+
+  return 0;
+}
+
+/**
+ * 
+ * @brief return initialized server_data
+ * @param src_info tensors info shared with serversrc element
+ * @return void* casted type of TensorQueryServerData*
+ */
+query_server_handle
+nnstreamer_query_server_data_new (void)
+{
+  TensorQueryServerData *sdata = g_new (TensorQueryServerData, 1);
```

Comment:
How about using g_try_new(... and adding some null checking after it?

Reply:
That looks better. Thank you